src/par_iter/collect/mod.rs

```diff
@@ -0,0 +1,38 @@
+use super::ExactParallelIterator;
+use std::isize;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+mod consumer;
+use self::consumer::CollectConsumer;
+
+pub fn collect_into<PAR_ITER,T>(mut pi: PAR_ITER, v: &mut Vec<T>)
+    where PAR_ITER: ExactParallelIterator<Item=T>,
+          PAR_ITER: ExactParallelIterator,
+          T: Send,
+{
+    let len = pi.len();
+    assert!(len < (isize::MAX) as usize);
+
+    v.truncate(0); // clear any old data
+    v.reserve(len); // reserve enough space
+    let target = v.as_mut_ptr(); // get a raw ptr
+    let writes = AtomicUsize::new(0);
+    let consumer = unsafe { // assert that target..target+len is unique and writable
+        CollectConsumer::new(&writes, target, len)
+    };
+    pi.drive(consumer);
```

Comment:
So, what I was saying about filter made no sense -- well, in particular, we couldn't adapt the approach you suggested to work with bounded parallel iterations, unless we added some sort of enumerated for that case which gave you the kind of "original index" (also in the case of a hand-written collect, I'd have to extend the protocol so that there is a version like `BoundedConsumer` where, for each split, you find out the bound at each step).